src/Env.hs

```diff
@@ -0,0 +1,255 @@
+import Debug.Trace
+import Data.Functor.Classes
+import Data.List
+import Data.Maybe (fromMaybe)
+import Data.Typeable
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import Text.Megaparsec.Expr
+--import Text.PrettyPrint
+--import Text.PrettyPrint.HughesPJClass hiding (colon, comma, parens, braces)
+import qualified Data.Map.Strict as M
+import qualified Text.Megaparsec.Char.Lexer as L
+
+----------------------------------- AST -----------------------------------------------
+
+data TypeName = TypeNameConst String -- these are all names, e.g. “Set”
+        | AllT -- specifically for global selection in Style
+    deriving (Show, Eq, Typeable)
+
+data TypeVar = TypeVarConst String
+     deriving (Show, Eq, Typeable, Ord)
+
+data Var = VarConst String
+     deriving (Show, Eq, Typeable, Ord)
+
+data Y = TypeVarY TypeVar | VarY Var
+     deriving (Show, Eq, Typeable, Ord)
+
+
+data T = TTypeVar TypeVar
+    | TConstr ConstructorInvoker
+    deriving (Show, Eq, Typeable)
+
+
+data ConstructorInvoker = ConstructorInvoker { nameCons :: String, argCons:: [Arg]}
+    deriving (Eq, Typeable)
+instance Show ConstructorInvoker where
+    show (ConstructorInvoker nameCons argCons) = nString ++ "(" ++ aString ++ ")"
+        where nString = show nameCons
+              aString = show argCons
+
+data Arg = AVar Var
+    | AT T
+    deriving (Show, Eq, Typeable)
+
+data K = Ktype Type
+     | KT T
+     deriving (Show, Eq, Typeable)
+
+data Type = TypeConst String
+     deriving (Show, Eq, Typeable)
+
+
+data Prop = PropConst String
+     deriving (Show, Eq, Typeable)
+
+
+----------------------------------- Parser --------------------------------------------
+
+typeNameParser ::Parser TypeName
+typeNameParser = do
+    i <- identifier
+    return (TypeNameConst i)
+
+typeParser :: Parser Type
+typeParser = do
+    rword "type"
+    return (TypeConst "type")
+
+varParser :: Parser Var
+varParser = do
+    i <- identifier
+    return (VarConst i)
+
+typeVarParser :: Parser TypeVar
+typeVarParser = do
+    aps
+    i <- identifier
+    return (TypeVarConst i)
+
+yParser, y1, y2 :: Parser Y
+yParser = try y1 <|> y2
+y1 = do
+  i <- varParser
+  return (VarY i)
+y2 = do
+  i <- typeVarParser
+  return (TypeVarY i)
+
+propParser :: Parser Prop
+propParser = do
+    rword "Prop"
+    return (PropConst "Prop")
+
+
+tParser, tConstructorInvokerParser, typeVarParser' :: Parser T
+tParser = try tConstructorInvokerParser <|> typeVarParser'
+tConstructorInvokerParser = do
+    i         <- identifier
+    arguments <- try (parens (argParser `sepBy1` comma)) <|> emptyArgList
+    return (TConstr (ConstructorInvoker {nameCons = i, argCons = arguments}))
+typeVarParser' = do
+    i <- typeVarParser
+    return (TTypeVar i)
+
+argParser, varParser', tParser'  :: Parser Arg
+argParser = try tParser' <|> varParser'
+varParser' = do
+     i <- varParser
+     return (AVar i)
+tParser' = do
+    t <- tParser
+    return (AT t)
+
+emptyArgList :: Parser [Arg]
```

Comment:
Any particular reason why you require `()` here? This breaks the existing syntax of `Set A` and requires `Set() A` instead, which causes most of the examples to fail.

Reply:
Yes, since in our grammar we might give arguments for type definitions.  For example, specify a vector space as an argument to a vector. I haven’t defined a syntactic sugar for that yet but we definitely should consider that.